File: server/src/main/com/broll/gainea/server/core/objects/Soldier.java
package com.broll.gainea.server.core.objects;

import com.broll.gainea.server.core.battle.BattleContext;
import com.broll.gainea.server.core.battle.FightingPower;
import com.broll.gainea.server.core.fractions.Fraction;
import com.broll.gainea.server.core.player.Player;

public class Soldier extends Unit {

    private Fraction fraction;
    private boolean commander;

    public Soldier(Player owner) {
        super(owner);
        if (owner != null) {
            this.fraction = owner.getFraction();
        }
    }

    public void setCommander(boolean commander) {
        this.commander = commander;
    }

    public boolean isCommander() {
        return commander;
    }

    @Override
    public FightingPower calcFightingPower(BattleContext context) {
        return fraction.calcFightingPower(this, context);
    }

    public Fraction getFraction() {
        return fraction;
    }
}

File: server/src/main/com/broll/gainea/server/core/objects/Soldier.kt
package com.broll.gainea.server.core.objects

import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.battle.BattleContext
import com.broll.gainea.server.core.battle.FightingPower
import com.broll.gainea.server.core.fractions.Fraction
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.player.isNeutral

open class Soldier(owner: Player,
                   var fraction: Fraction? = null,
                   var isCommander: Boolean = false) : Unit(owner) {

    init {
        if (owner.isNeutral()) {
            this.fraction = null
        }
    }

    override fun init(game: Game) {
        super.init(game)
        if (isCommander) {
            description = "Feldherr"
        }
    }

    override fun calcFightingPower(context: BattleContext): FightingPower {
        return if (fraction == null) {
            super.calcFightingPower(context)
        } else fraction!!.calcFightingPower(this, context)
    }
}
